### PR TITLE
Dataclasses: support mutable properties

### LINKED ISSUE #6493: Internal error on dataclasses with properties
On CPython 3.7.2 [GCC 7.3.0] on linux, with mypy 0.670

When attempting to follow the guidelines described in [this article](https://blog.florimondmanca.com/reconciling-dataclasses-and-properties-in-python)

As illustrated here:

```
from dataclasses import dataclass
from typing import Optional


@dataclass(init=False)
class C:
    a: Optional[int]

    def __init__(self, a: Optional[int] = None) -> None:
        self._a = None
        self.a = a

    @property
    def a(self) -> Optional[int]:
        return self._a

    @a.setter
    def a(self, value: Optional[int]) -> Optional[int]:
        self._a = value
        return self._a
```

Mypy (without options) raises an `INTERNAL ERROR` with the following:

```
Traceback (most recent call last):
  File ".../bin/mypy", line 10, in <module>
    sys.exit(console_entry())
  File ".../lib/python3.7/site-packages/mypy/__main__.py", line 7, in console_entry
    main(None)
  File ".../lib/python3.7/site-packages/mypy/main.py", line 91, in main
    res = build.build(sources, options, None, flush_errors, fscache)
  File ".../lib/python3.7/site-packages/mypy/build.py", line 162, in build
    result = _build(sources, options, alt_lib_path, flush_errors, fscache)
  File ".../lib/python3.7/site-packages/mypy/build.py", line 217, in _build
    graph = dispatch(sources, manager)
  File ".../lib/python3.7/site-packages/mypy/build.py", line 2360, in dispatch
    process_graph(graph, manager)
  File ".../lib/python3.7/site-packages/mypy/build.py", line 2660, in process_graph
    process_stale_scc(graph, scc, manager)
  File ".../lib/python3.7/site-packages/mypy/build.py", line 2761, in process_stale_scc
    graph[id].semantic_analysis()
  File ".../lib/python3.7/site-packages/mypy/build.py", line 1896, in semantic_analysis
    self.manager.semantic_analyzer.visit_file(self.tree, self.xpath, self.options, patches)
  File ".../lib/python3.7/site-packages/mypy/semanal.py", line 306, in visit_file
    self.accept(d)
  File ".../lib/python3.7/site-packages/mypy/semanal.py", line 3787, in accept
    node.accept(self)
  File ".../lib/python3.7/site-packages/mypy/nodes.py", line 846, in accept
    return visitor.visit_class_def(self)
  File ".../lib/python3.7/site-packages/mypy/semanal.py", line 804, in visit_class_def
    self.analyze_class(defn)
  File ".../lib/python3.7/site-packages/mypy/semanal.py", line 822, in analyze_class
    self.analyze_class_body_common(defn)
  File ".../lib/python3.7/site-packages/mypy/semanal.py", line 830, in analyze_class_body_common
    self.apply_class_plugin_hooks(defn)
  File ".../lib/python3.7/site-packages/mypy/semanal.py", line 892, in apply_class_plugin_hooks
    hook(ClassDefContext(defn, decorator, self))
  File ".../lib/python3.7/site-packages/mypy/plugins/dataclasses.py", line 312, in dataclass_class_maker_callback
    transformer.transform()
  File ".../lib/python3.7/site-packages/mypy/plugins/dataclasses.py", line 79, in transform
    attributes = self.collect_attributes()
  File ".../lib/python3.7/site-packages/mypy/plugins/dataclasses.py", line 189, in collect_attributes
    assert isinstance(node, Var)
AssertionError:
```
Any suggestions about workarounds or better implementations of the initial premise would be much appreciated!

### PR DESCRIPTION
**Pull request by ckanesan**
Skip over attributes that are redefined as properties on dataclass construction.

As far as I understand, not redefining a symbol within the same scope is a fundamental assumption in mypy. However, I have not been able to figure out another way to get a mutable property that is also part of the `dataclasses.fields`. It seems that we are falling within the cracks in the sense that the property methods actually mask the attribute with the desired functionality and all seems to go well in the end. 

Any indications on how to make a cleaner implementation would be most welcome!

Fixes #6493

### CODE REVIEW

File: mypy/plugins/dataclasses.py [hunk 1]
```diff
@@ -204,7 +204,10 @@ def collect_attributes(self) -> List[DataclassAttribute]:
                 continue
 
             node = cls.info.names[lhs.name].node
-            assert isinstance(node, Var)
+            assert isinstance(node, (Var, OverloadedFuncDef))
+
+            if isinstance(node, OverloadedFuncDef):
+                continue
```
Comment (ilevkivskyi) [inline mypy/plugins/dataclasses.py]: This is still dangerous, what if `node` is a `Decorator`? I would just replace the old assert with `if not isinstance(node, Var): continue`.

**Comment by ckanesan** [comment]
@ilevkivskyi Thanks for the suggestion. I think you are right but I can't seem to wrap my head around the possible edge cases.

**Comment by ilevkivskyi** [comment]
Thanks for update! Unfortunately, we will need to wait with this PR until  #6432 is fixed.

**Comment by ilevkivskyi** [comment]
This PR may be unblocked now. Could you please rebase on current master?

Also it looks like a fix identical to the one you add may also fix https://github.com/python/mypy/issues/6715 would you like to add it to this PR? Thank you!

**Comment by ckanesan** [comment]
@ilevkivskyi Sorry for the delay, I was absent for a while due to personal obligations. Is this still okay to merge?

**Comment by ilevkivskyi** [comment]
@ckanesan No problem. Please note there is still a (legitimate) failure on the new semantic analyzer. So you should probably update the test to only run on new analyzer, or put a type ignore on that line.

**Comment by ilevkivskyi** [comment]
Also as I mentioned there is almost identical issue for attrs #6715. It would be great if you can fix it too.

**Comment by 97littleleaf11** [comment]
Jukka closed the original issue, https://github.com/python/mypy/issues/6493#issuecomment-586349590
> This no longer crashes, so I'm closing this issue. Mypy still doesn't support dataclasses with properties, however -- feel free to create a separate issue about this feature.

I am going to close this PR since it's inactive. Feel free to open a new one at your convenience!

**Pull request closed without merging**